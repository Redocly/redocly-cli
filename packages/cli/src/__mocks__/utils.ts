export const getFallbackEntryPointsOrExit = jest.fn((entrypoints) => entrypoints.map(() => ({ path: '' })));
export const getTotals = jest.fn(() => ({ errors: 0 }));
export const dumpBundle = jest.fn(() => '');
export const slash = jest.fn();
export const pluralize = jest.fn();
export const getExecutionTime = jest.fn();
export const printExecutionTime = jest.fn();
export const printUnusedWarnings = jest.fn();
export const printLintTotals = jest.fn();
export const getOutputFileName = jest.fn(() => ({ outputFile: 'test.yaml', ext: 'yaml' }));
export const handleError = jest.fn();
