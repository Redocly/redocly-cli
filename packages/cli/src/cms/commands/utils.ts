import { pause } from '@redocly/openapi-core';

/**
 * This function retries an operation until a condition is met or a timeout is exceeded.
 * If the condition is not met within the timeout, an error is thrown.
 * @operation The operation to retry.
 * @condition The condition to check after each operation result. Return false to continue retrying. Return true to stop retrying.
 *            If not provided, the first result will be returned.
 * @param onConditionNotMet Will be called with the last result right after checking condition and before timeout and retrying.
 * @param onRetry Will be called right before retrying operation with the last result before retrying.
 * @param startTime The start time of the operation. Default is the current time.
 * @param retryTimeoutMs The maximum time to retry the operation. Default is 10 minutes.
 * @param retryIntervalMs The interval between retries. Default is 5 seconds.
 */
export async function retryUntilConditionMet<T>({
  operation,
  condition,
  onConditionNotMet,
  onRetry,
  startTime = Date.now(),
  retryTimeoutMs = 600000, // 10 min
  retryIntervalMs = 5000, // 5 sec
}: {
  operation: () => Promise<T>;
  condition?: ((result: T) => boolean) | null;
  onConditionNotMet?: (lastResult: T) => void;
  onRetry?: (lastResult: T) => void | Promise<void>;
  startTime?: number;
  retryTimeoutMs?: number;
  retryIntervalMs?: number;
}): Promise<T> {
  async function attempt(): Promise<T> {
    const result = await operation();

    if (!condition) {
      return result;
    }

    if (condition(result)) {
      return result;
    } else if (Date.now() - startTime > retryTimeoutMs) {
      throw new Error('Timeout exceeded');
    } else {
      onConditionNotMet?.(result);
      await pause(retryIntervalMs);
      await onRetry?.(result);
      return attempt();
    }
  }

  return attempt();
}
