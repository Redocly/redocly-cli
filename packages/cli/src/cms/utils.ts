export class DeploymentError extends Error {}
