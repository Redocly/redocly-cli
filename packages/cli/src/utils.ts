import { basename, dirname, extname, join, resolve } from 'path';
import { blue, gray, green, red, yellow } from 'colorette';
import { performance } from "perf_hooks";
import * as glob from 'glob-promise';
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import { Writable } from 'stream';
import {
  BundleOutputFormat,
  Config,
  LintConfig,
  ResolveError,
  YamlParseError,
  parseYaml,
  stringifyYaml
} from '@redocly/openapi-core';
import { Totals, outputExtensions, Entrypoint } from './types';

export async function getFallbackEntryPointsOrExit(argsEntrypoints: string[] | undefined, config: Config): Promise<Entrypoint[]> {
  const  { apis } = config;
  const shouldFallbackToAllDefinitions = !isNotEmptyArray(argsEntrypoints) && apis && Object.keys(apis).length > 0;
  const res = shouldFallbackToAllDefinitions
    ? Object.entries(apis).map(([alias, { root }]) => ({
        path: resolve(getConfigDirectory(config), root),
        alias,
      }))
    : (await expandGlobsInEntrypoints(argsEntrypoints!, config));
  if (!isNotEmptyArray(res)) {
    process.stderr.write('error: missing required argument `entrypoints`.\n');
    process.exit(1);
  }
  return res;
}

function getConfigDirectory(config: Config) {
  return config.configFile ? dirname(config.configFile) : process.cwd();
}

function isNotEmptyArray<T>(args?: T[]): boolean {
  return Array.isArray(args) && !!args.length;
}

function getAliasOrPath(config: Config, aliasOrPath: string): Entrypoint {
  return config.apis[aliasOrPath]
    ? { path: config.apis[aliasOrPath]?.root, alias: aliasOrPath }
    : { path: aliasOrPath };
}

async function expandGlobsInEntrypoints(args: string[], config: Config) {
  return (await Promise.all((args as string[]).map(async aliasOrPath => {
    return glob.hasMagic(aliasOrPath)
      ? (await glob(aliasOrPath)).map((g: string) => getAliasOrPath(config, g))
      : getAliasOrPath(config, aliasOrPath);
  }))).flat();
}

export function getExecutionTime(startedAt: number) {
  return process.env.NODE_ENV === 'test'
    ? '<test>ms'
    : `${Math.ceil(performance.now() - startedAt)}ms`;
}

export function printExecutionTime(commandName: string, startedAt: number, entrypoint: string) {
  const elapsed = getExecutionTime(startedAt);
  process.stderr.write(gray(`\n${entrypoint}: ${commandName} processed in ${elapsed}\n\n`));
}

export function pathToFilename(path: string, pathSeparator: string) {
  return path
    .replace(/~1/g, '/')
    .replace(/~0/g, '~')
    .replace(/^\//, '')
    .replace(/\//g, pathSeparator);
}

export class CircularJSONNotSupportedError extends Error {
  constructor(public originalError: Error) {
    super(originalError.message);
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, CircularJSONNotSupportedError.prototype);
  }
}

export function dumpBundle(obj: any, format: BundleOutputFormat, dereference?: boolean): string {
  if (format === 'json') {
    try {
      return JSON.stringify(obj, null, 2);
    } catch (e) {
      if (e.message.indexOf('circular') > -1) {
        throw new CircularJSONNotSupportedError(e);
      }
      throw e;
    }
  } else {
    return stringifyYaml(obj, {
      noRefs: !dereference,
    });
  }
}

export function saveBundle(filename: string, output: string) {
  fs.mkdirSync(path.dirname(filename), { recursive: true });
  fs.writeFileSync(filename, output);
}

export async function promptUser(query: string, hideUserInput = false): Promise<string> {
  return new Promise((resolve) => {
    let output: Writable = process.stdout;
    let isOutputMuted = false;

    if (hideUserInput) {
      output = new Writable({
        write: (chunk, encoding, callback) => {
          if (!isOutputMuted) {
            process.stdout.write(chunk, encoding);
          }
          callback();
        },
      });
    }

    const rl = readline.createInterface({
      input: process.stdin,
      output,
      terminal: true,
      historySize: hideUserInput ? 0 : 30,
    });

    rl.question(`${query}:\n\n  `, (answer) => {
      rl.close();
      resolve(answer);
    });

    isOutputMuted = hideUserInput;
  });
}

export function readYaml(filename: string) {
  return parseYaml(fs.readFileSync(filename, 'utf-8'), { filename });
}

export function writeYaml(data: any, filename: string, noRefs = false) {
  const content = stringifyYaml(data, { noRefs });

  if (process.env.NODE_ENV === 'test') {
    process.stderr.write(content);
    return;
  }
  fs.writeFileSync(filename, content);
}

export function pluralize(label: string, num: number) {
  if (label.endsWith('is')) {
    [label] = label.split(' ');
    return num === 1 ? `${label} is` : `${label}s are`;
  }
  return num === 1 ? `${label}` : `${label}s`;
}

export function handleError(e: Error, ref: string) {
  if (e instanceof ResolveError) {
    process.stderr.write(
      `Failed to resolve entrypoint definition at ${ref}:\n\n  - ${e.message}.\n\n`,
    );
  } else if (e instanceof YamlParseError) {
    process.stderr.write(
      `Failed to parse entrypoint definition at ${ref}:\n\n  - ${e.message}.\n\n`,
    );
    // TODO: codeframe
  } else { // @ts-ignore
    if (e instanceof CircularJSONNotSupportedError) {
      process.stderr.write(
        red(`Detected circular reference which can't be converted to JSON.\n`) +
        `Try to use ${blue('yaml')} output or remove ${blue('--dereferenced')}.\n\n`,
      );
    } else {
      process.stderr.write(`Something went wrong when processing ${ref}:\n\n  - ${e.message}.\n\n`);
      throw e;
    }
  }
}

export function printLintTotals(totals: Totals, definitionsCount: number) {
  const ignored = totals.ignored
    ? yellow(`${totals.ignored} ${pluralize('problem is', totals.ignored)} explicitly ignored.\n\n`)
    : '';

  if (totals.errors > 0) {
    process.stderr.write(
      red(
        `❌ Validation failed with ${totals.errors} ${pluralize('error', totals.errors)}${
          totals.warnings > 0
            ? ` and ${totals.warnings} ${pluralize('warning', totals.warnings)}`
            : ''
        }.\n${ignored}`,
      ),
    );
  } else if (totals.warnings > 0) {
    process.stderr.write(
      green(`Woohoo! Your OpenAPI ${pluralize('definition is', definitionsCount)} valid. 🎉\n`),
    );
    process.stderr.write(
      yellow(`You have ${totals.warnings} ${pluralize('warning', totals.warnings)}.\n${ignored}`),
    );
  } else {
    process.stderr.write(
      green(
        `Woohoo! Your OpenAPI ${pluralize(
          'definition is',
          definitionsCount,
        )} valid. 🎉\n${ignored}`,
      ),
    );
  }

  if (totals.errors > 0) {
    process.stderr.write(
      gray(`run \`openapi lint --generate-ignore-file\` to add all problems to the ignore file.\n`),
    );
  }

  process.stderr.write('\n');
}

export function getOutputFileName(
  entrypoint: string,
  entries: number,
  output?: string,
  ext?: BundleOutputFormat,
) {
  if (!output) {
    return { outputFile: 'stdout', ext: ext || 'yaml' };
  }

  let outputFile = output;
  if (entries > 1) {
    ext = ext || (extname(entrypoint).substring(1) as BundleOutputFormat);
    if (!outputExtensions.includes(ext as any)) {
      throw new Error(`Invalid file extension: ${ext}.`);
    }
    outputFile = join(output, basename(entrypoint, extname(entrypoint))) + '.' + ext;
  } else {
    if (output) {
      ext = ext || (extname(output).substring(1) as BundleOutputFormat);
    }
    ext = ext || (extname(entrypoint).substring(1) as BundleOutputFormat);
    if (!outputExtensions.includes(ext as any)) {
      throw new Error(`Invalid file extension: ${ext}.`);
    }
    outputFile = join(dirname(outputFile), basename(outputFile, extname(outputFile))) + '.' + ext;
  }
  return { outputFile, ext };
}

export function printUnusedWarnings(config: LintConfig) {
  const { preprocessors, rules, decorators } = config.getUnusedRules();
  if (rules.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused rules found in ${blue(config.configFile || '')}: ${rules.join(', ')}.\n`,
      ),
    );
  }
  if (preprocessors.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused preprocessors found in ${blue(
          config.configFile || '',
        )}: ${preprocessors.join(', ')}.\n`,
      ),
    );
  }
  if (decorators.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused decorators found in ${blue(config.configFile || '')}: ${decorators.join(
          ', ',
        )}.\n`,
      ),
    );
  }

  if (rules.length || preprocessors.length) {
    process.stderr.write(`Check the spelling and verify you added plugin prefix.\n`);
  }
}

export function exitWithError(message: string) {
  process.stderr.write(red(message)+ '\n\n');
  process.exit(1);
}

/**
 * Checks if dir is subdir of parent
 */
export function isSubdir(parent: string, dir: string): boolean {
  const relative = path.relative(parent, dir);
  return !!relative && !/^..($|\/)/.test(relative) && !path.isAbsolute(relative);
}
