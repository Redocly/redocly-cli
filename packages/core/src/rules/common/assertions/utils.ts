import { isRef } from '../../../ref-utils';
import { Problem, ProblemSeverity, UserContext } from '../../../walk';
import { asserts } from './asserts';

export type OrderDirection = 'asc' | 'desc';

export type OrderOptions = {
  direction: OrderDirection;
  property: string;
};

export type AssertToApply = {
  name: string;
  assertId?: string;
  conditions: any;
  message?: string;
  severity?: ProblemSeverity;
  suggest?: string[];
  runsOnKeys: boolean;
  runsOnValues: boolean;
};

export function buildVisitorObject(
  subject: string,
  context: Record<string, any>[],
  subjectVisitor: any,
) {
  if (!context) {
    return { [subject]: subjectVisitor };
  }

  let currentVisitorLevel: Record<string, any> = {};
  const visitor: Record<string, any> = currentVisitorLevel;

  for (let index = 0; index < context.length; index++) {
    const node = context[index];
    if (context.length === index + 1 && node.type === subject) {
      // Visitors don't work properly for the same type nested nodes, so
      // as a workaround for that we don't create separate visitor for the last element
      // which is the same as subject;
      // we will check includes/excludes it in the last visitor.
      continue;
    }
    const matchParentKeys = node.matchParentKeys;
    const excludeParentKeys = node.excludeParentKeys;

    if (matchParentKeys && excludeParentKeys) {
      throw new Error(
        `Both 'matchParentKeys' and 'excludeParentKeys' can't be under one context item`,
      );
    }

    if (matchParentKeys || excludeParentKeys) {
      currentVisitorLevel[node.type] = {
        skip: (_value: any, key: string) => {
          if (matchParentKeys) {
            return !matchParentKeys.includes(key);
          }
          if (excludeParentKeys) {
            return excludeParentKeys.includes(key);
          }
        },
      };
    } else {
      currentVisitorLevel[node.type] = {};
    }
    currentVisitorLevel = currentVisitorLevel[node.type];
  }

  currentVisitorLevel[subject] = subjectVisitor;

  return visitor;
}

export function buildSubjectVisitor(
  properties: string | string[],
  asserts: AssertToApply[],
  context?: Record<string, any>[],
) {
  return function (node: any, { report, location, key, type, resolve }: UserContext) {
    // We need to check context's last node if it has the same type as subject node;
    // if yes - that means we didn't create context's last node visitor,
    // so we need to handle 'matchParentKeys' and 'excludeParentKeys' conditions here;
    if (context) {
      const lastContextNode = context[context.length - 1];
      if (lastContextNode.type === type.name) {
        const matchParentKeys = lastContextNode.matchParentKeys;
        const excludeParentKeys = lastContextNode.excludeParentKeys;

        if (matchParentKeys && !matchParentKeys.includes(key)) {
          return;
        }
        if (excludeParentKeys && excludeParentKeys.includes(key)) {
          return;
        }
      }
    }

    if (properties) {
      properties = Array.isArray(properties) ? properties : [properties];
    }

    for (const assert of asserts) {
      if (properties) {
        for (const property of properties) {
          // we can have resolvable scalar so need to resolve value here.
          const value = isRef(node[property]) ? resolve(node[property])?.node : node[property];
          runAssertion(value, assert, location.child(property), report);
        }
      } else {
        runAssertion(Object.keys(node), assert, location.key(), report);
      }
    }
  };
}

export function getIntersectionLength(keys: string[], properties: string[]): number {
  const props = new Set(properties);
  let count = 0;
  for (const key of keys) {
    if (props.has(key)) {
      count++;
    }
  }
  return count;
}

export function isOrdered(value: any[], options: OrderOptions | OrderDirection): boolean {
  const direction = (options as OrderOptions).direction || (options as OrderDirection);
  const property = (options as OrderOptions).property;
  for (let i = 1; i < value.length; i++) {
    let currValue = value[i];
    let prevVal = value[i - 1];

    if (property) {
      if (!value[i][property] || !value[i - 1][property]) {
        return false; // property doesn't exist, so collection is not ordered
      }
      currValue = value[i][property];
      prevVal = value[i - 1][property];
    }

    const result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
    if (!result) {
      return false;
    }
  }
  return true;
}

function runAssertion(
  values: string | string[],
  assert: AssertToApply,
  location: any,
  report: (problem: Problem) => void,
) {
  const lintResult = asserts[assert.name](values, assert.conditions);
  if (!lintResult) {
    report({
      message: assert.message || `The ${assert.assertId} doesn't meet required conditions`,
      location,
      forceSeverity: assert.severity,
      suggest: assert.suggest,
      ruleId: assert.assertId,
    });
  }
}
