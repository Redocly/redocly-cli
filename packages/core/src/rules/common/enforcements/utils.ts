import { ProblemSeverity, UserContext } from '../../../walk';
import { rules as genericRules } from './generic-rules';

export type OrderDirection = 'asc' | 'desc';

export type OrderOptions = {
  direction: OrderDirection;
  property?: string;
};

export type Rule = {
  name: string;
  conditions: any;
  description: string;
  severity: ProblemSeverity;
  runsOnKeys: boolean;
  runsOnValues: boolean;
};

export const ALL_KEYS = '$keys';

/** Sets the value at path of object. If a portion of path doesn't exist, it's created.  */
export const objectSet = (path: string[], value: any) => {
  return path.reverse().reduce((acc, key) => ({[key]: acc}), value);
}

export const formRule = (lastNodeName: string, propsToRules:  {[key: string]: Rule[]}) => {
  return {
    [lastNodeName]: function(node: any, { report, location }: UserContext) {
      for (const prop of Object.keys(propsToRules)) {
        for (const rule of propsToRules[prop]) {
          const value = prop === ALL_KEYS ? node : node[prop];
          const lintResult = (genericRules as {[key: string]: any})[rule.name](value, rule.conditions);
          if (!lintResult) {
            report({
              message: rule.description,
              location: prop === ALL_KEYS ? location.key() : location.child(prop).key(),
              forceSeverity: rule.severity
            });
          }
        }
      }
    }
  }
}

export const getCounts = (node: any, properties: string[]): number => {
  let counter = 0;
  for (const prop of Object.keys(node)) {
    if (properties.includes(prop)) {
      counter++;
    }
  }
  return counter;
}

export const isOrdered = (value: any[], options: OrderOptions | OrderDirection): boolean => {
  const direction = (options as OrderOptions).direction || options as OrderDirection;
  const property = (options as OrderOptions).property;
  let result = true;
  for (let i=1; i<value.length; i++) {
    let currValue = value[i];
    let prevVal = value[i-1];

    if (property) {
      if (!value[i][property] || !value[i-1][property]) {
        return false; // property doesn't exist, so collection is not ordered
      }
      currValue = value[i][property];
      prevVal = value[i-1][property]
    }

    result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
    if (!result) {
      break;
    }
  }
  return result;
}