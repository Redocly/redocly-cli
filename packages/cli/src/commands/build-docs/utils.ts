import { createElement } from 'react';
import { createStore, Redoc } from 'redoc';
import { Config, isAbsoluteUrl } from '@redocly/openapi-core';

import { renderToString } from 'react-dom/server';
import { ServerStyleSheet } from 'styled-components';
import { compile } from 'handlebars';
import { join } from 'path';
import { existsSync, lstatSync, readFileSync } from 'fs';

import type { BuildDocsOptions } from './types';
import { red } from 'colorette';
import { exitWithError } from '../../utils';

export function getObjectOrJSON(
  openapiOptions: string | Record<string, unknown>,
  config: Config
): JSON | Record<string, unknown> | Config {
  switch (typeof openapiOptions) {
    case 'object':
      return openapiOptions;
    case 'string':
      try {
        if (existsSync(openapiOptions) && lstatSync(openapiOptions).isFile()) {
          return JSON.parse(readFileSync(openapiOptions, 'utf-8'));
        } else {
          return JSON.parse(openapiOptions);
        }
      } catch (e) {
        process.stderr.write(
          red(
            `Encountered error:\n\n${openapiOptions}\n\nis neither a file with a valid JSON object neither a stringified JSON object.`
          )
        );
        exitWithError(e);
      }
      break;
    default: {
      if (config) {
        process.stderr.write(`Found ${config.configFile} and using theme.openapi options`);

        return config.theme.openapi ? config.theme.openapi : {};
      }
      return {};
    }
  }
  return {};
}

export async function getPageHTML(
  api: any,
  pathToApi: string,
  {
    cdn,
    title,
    disableGoogleFont,
    templateFileName,
    templateOptions,
    redocOptions = {},
    redocCurrentVersion,
  }: BuildDocsOptions
) {
  process.stderr.write('Prerendering docs');

  const apiUrl = redocOptions.specUrl || (isAbsoluteUrl(pathToApi) ? pathToApi : undefined);
  const store = await createStore(api, apiUrl, redocOptions);
  const sheet = new ServerStyleSheet();

  const html = renderToString(sheet.collectStyles(createElement(Redoc, { store })));
  const state = await store.toJS();
  const css = sheet.getStyleTags();

  templateFileName = templateFileName ? templateFileName : join(__dirname, './template.hbs');
  const template = compile(readFileSync(templateFileName).toString());
  return template({
    redocHTML: `
      <div id="redoc">${html || ''}</div>
      <script>
      ${`const __redoc_state = ${sanitizeJSONString(JSON.stringify(state))};` || ''}

      var container = document.getElementById('redoc');
      Redoc.${'hydrate(__redoc_state, container)'};

      </script>`,
    redocHead:
      (cdn
        ? '<script src="https://cdn.redoc.ly/redoc/latest/bundles/redoc.standalone.js"></script>'
        : `<script src="https://cdn.redoc.ly/redoc/v${redocCurrentVersion}/bundles/redoc.standalone.js"></script>`) +
      css,
    title: title || api.info.title || 'ReDoc documentation',
    disableGoogleFont,
    templateOptions,
  });
}

export function sanitizeJSONString(str: string): string {
  return escapeClosingScriptTag(escapeUnicode(str));
}

// see http://www.thespanner.co.uk/2011/07/25/the-json-specification-is-now-wrong/
export function escapeClosingScriptTag(str: string): string {
  return str.replace(/<\/script>/g, '<\\/script>');
}

// see http://www.thespanner.co.uk/2011/07/25/the-json-specification-is-now-wrong/
export function escapeUnicode(str: string): string {
  return str.replace(/\u2028|\u2029/g, (m) => '\\u202' + (m === '\u2028' ? '8' : '9'));
}
