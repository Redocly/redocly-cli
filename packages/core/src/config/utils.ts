import { green, yellow } from 'colorette';
import { assignExisting } from '../utils';
import { Config } from './config';

import type {
  Api,
  DeprecatedRawConfig,
  Plugin,
  RawConfig,
  RawResolveConfig,
  ResolveConfig,
  ResolvedLintConfig,
  RulesFields,
} from './types';

export function parsePresetName(presetName: string): { pluginId: string; configName: string } {
  if (presetName.indexOf('/') > -1) {
    const [pluginId, configName] = presetName.split('/');
    return { pluginId, configName };
  } else {
    return { pluginId: '', configName: presetName };
  }
}

export function transformApiDefinitionsToApis(
  apiDefinitions: Record<string, string> = {},
): Record<string, Api> {
  let apis: Record<string, Api> = {};
  for (const [apiName, apiPath] of Object.entries(apiDefinitions)) {
    apis[apiName] = { root: apiPath };
  }
  return apis;
}

export function prefixRules<T extends Record<string, any>>(rules: T, prefix: string) {
  if (!prefix) return rules;

  const res: any = {};
  for (const name of Object.keys(rules)) {
    res[`${prefix}/${name}`] = rules[name];
  }

  return res;
}

export function mergeExtends(rulesConfList: ResolvedLintConfig[]) {
  const result: Omit<ResolvedLintConfig, RulesFields> &
    Required<Pick<ResolvedLintConfig, RulesFields>> = {
    rules: {},
    oas2Rules: {},
    oas3_0Rules: {},
    oas3_1Rules: {},

    preprocessors: {},
    oas2Preprocessors: {},
    oas3_0Preprocessors: {},
    oas3_1Preprocessors: {},

    decorators: {},
    oas2Decorators: {},
    oas3_0Decorators: {},
    oas3_1Decorators: {},

    plugins: [],
    pluginPaths: [],
    extendPaths: [],
  };

  for (let rulesConf of rulesConfList) {
    if (rulesConf.extends) {
      throw new Error(
        `\`extends\` is not supported in shared configs yet: ${JSON.stringify(
          rulesConf,
          null,
          2,
        )}.`,
      );
    }

    Object.assign(result.rules, rulesConf.rules);
    Object.assign(result.oas2Rules, rulesConf.oas2Rules);
    assignExisting(result.oas2Rules, rulesConf.rules || {});
    Object.assign(result.oas3_0Rules, rulesConf.oas3_0Rules);
    assignExisting(result.oas3_0Rules, rulesConf.rules || {});
    Object.assign(result.oas3_1Rules, rulesConf.oas3_1Rules);
    assignExisting(result.oas3_1Rules, rulesConf.rules || {});

    Object.assign(result.preprocessors, rulesConf.preprocessors);
    Object.assign(result.oas2Preprocessors, rulesConf.oas2Preprocessors);
    assignExisting(result.oas2Preprocessors, rulesConf.preprocessors || {});
    Object.assign(result.oas3_0Preprocessors, rulesConf.oas3_0Preprocessors);
    assignExisting(result.oas3_0Preprocessors, rulesConf.preprocessors || {});
    Object.assign(result.oas3_1Preprocessors, rulesConf.oas3_1Preprocessors);
    assignExisting(result.oas3_1Preprocessors, rulesConf.preprocessors || {});

    Object.assign(result.decorators, rulesConf.decorators);
    Object.assign(result.oas2Decorators, rulesConf.oas2Decorators);
    assignExisting(result.oas2Decorators, rulesConf.decorators || {});
    Object.assign(result.oas3_0Decorators, rulesConf.oas3_0Decorators);
    assignExisting(result.oas3_0Decorators, rulesConf.decorators || {});
    Object.assign(result.oas3_1Decorators, rulesConf.oas3_1Decorators);
    assignExisting(result.oas3_1Decorators, rulesConf.decorators || {});

    result.plugins!.push(...(rulesConf.plugins || []));
    result.pluginPaths!.push(...(rulesConf.pluginPaths || []));
    result.extendPaths!.push(...new Set(rulesConf.extendPaths));
  }

  return result;
}

export function getMergedConfig(config: Config, entrypointAlias?: string): Config {
  const extendPaths = [
    ...Object.values(config.apis).map((api) => api?.lint?.extendPaths),
    config.rawConfig?.lint?.extendPaths,
  ]
    .flat()
    .filter(Boolean) as string[];

  const pluginPaths = [
    ...Object.values(config.apis).map((api) => api?.lint?.pluginPaths),
    config.rawConfig?.lint?.pluginPaths,
  ]
    .flat()
    .filter(Boolean) as string[];

  return entrypointAlias
    ? new Config(
        {
          ...config.rawConfig,
          lint: {
            ...(config.apis[entrypointAlias]
              ? config.apis[entrypointAlias].lint
              : config.rawConfig.lint),
            extendPaths,
            pluginPaths,
          },
          'features.openapi': {
            ...config['features.openapi'],
            ...config.apis[entrypointAlias]?.['features.openapi'],
          },
          'features.mockServer': {
            ...config['features.mockServer'],
            ...config.apis[entrypointAlias]?.['features.mockServer'],
          },
          // TODO: merge everything else here
        },
        config.configFile,
      )
    : config;
}

export function transformConfig(rawConfig: DeprecatedRawConfig | RawConfig): RawConfig {
  if ((rawConfig as RawConfig).apis && (rawConfig as DeprecatedRawConfig).apiDefinitions) {
    throw new Error("Do not use 'apiDefinitions' field. Use 'apis' instead.\n");
  }
  if (
    (rawConfig as RawConfig)['features.openapi'] &&
    (rawConfig as DeprecatedRawConfig).referenceDocs
  ) {
    throw new Error("Do not use 'referenceDocs' field. Use 'features.openapi' instead.\n");
  }
  const { apiDefinitions, referenceDocs, ...rest } = rawConfig as DeprecatedRawConfig & RawConfig;
  if (apiDefinitions) {
    process.stderr.write(
      `The ${yellow('apiDefinitions')} field is deprecated. Use ${green(
        'apis',
      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
    );
  }
  if (referenceDocs) {
    process.stderr.write(
      `The ${yellow('referenceDocs')} field is deprecated. Use ${green(
        'features.openapi',
      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
    );
  }
  return {
    'features.openapi': referenceDocs,
    apis: transformApiDefinitionsToApis(apiDefinitions),
    ...rest,
  };
}

export function getResolveConfig(resolve?: RawResolveConfig): ResolveConfig {
  return {
    http: {
      headers: resolve?.http?.headers ?? [],
      customFetch: undefined,
    },
  };
}

export function getUniquePlugins(plugins: Plugin[]): Plugin[] {
  const seen = new Set();
  const results = [];
  for (const p of plugins) {
    if (!seen.has(p.id)) {
      results.push(p);
      seen.add(p.id);
    } else if (p.id) {
      process.stderr.write(
        `Duplicate plugin id "${yellow(p.id)}".\n`,
      );
    }
  }
  return results;
}
