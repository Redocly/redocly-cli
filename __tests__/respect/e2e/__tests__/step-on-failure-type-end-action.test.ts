import { getParams, getCommandOutput, getFixturePath } from './utils';

// Snapshot should have two workflows, and first workflow should run only first step
it('should end workflow execution, context returns to the caller with applicable outputs, when step fails and onFailure action is of type `end`', () => {
  process.env.AUTH_TOKEN = 'Basic Og==';

  const params = getParams('../../lib-internal/cli.js', [
    'run',
    getFixturePath('step-on-failure-type-end-action.yaml'),
  ]);

  const result = getCommandOutput(params);

  expect(result).toMatchSnapshot();
  delete process.env.AUTH_TOKEN;
});
