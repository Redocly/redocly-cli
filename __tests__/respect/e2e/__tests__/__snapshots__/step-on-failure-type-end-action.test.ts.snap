// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should end workflow execution, context returns to the caller with applicable outputs, when step fails and onFailure action is of type \`end\` 1`] = `
"────────────────────────────────────────────────────────────────────────────────[0m
[0m
  Running workflow step-on-failure-type-end-action.yaml / events-crud[0m
[0m
  ✗ GET /special-events - step list-events[0m
    ✗ success criteria check
    ✓ status code check (Response code 200 matches one of description codes: [200, 400, 404])
    ✓ content-type check
    ✓ schema check
[0m
────────────────────────────────────────────────────────────────────────────────[0m
[0m
  Running workflow step-on-failure-type-end-action.yaml / get-museum-hours[0m
[0m
  ✓ GET /museum-hours - step get-museum-hours[0m
    ✓ success criteria check
    ✓ status code check (Response code 200 matches one of description codes: [200, 400, 404])
    ✓ content-type check
    ✓ schema check
[0m
[0m
  Failed tests info:
[0m
  Workflow name: events-crud[0m
[0m
    stepId - list-events[0m
    ✗ success criteria check[0m
      Checking simple criteria: {"condition":"$statusCode == 201"}[0m
      [0m
  Summary for step-on-failure-type-end-action.yaml
  
  Workflows: 1 passed, 1 failed, 2 total
  Steps: 2 passed, 1 failed, 3 total
  Checks: 7 passed, 1 failed, 8 total
  Time: <test>ms[0m
[0m
[0m
┌──────────────────────────────────────────────────────────────────────────────┬────────────┬─────────┬─────────┬──────────┬─────────┐
│ Filename                                                                     │ Workflows  │ Passed  │ Failed  │ Warnings │ Skipped │
├──────────────────────────────────────────────────────────────────────────────┼────────────┼─────────┼─────────┼──────────┼─────────┤
│ x step-on-failure-type-end-action.yaml                                       │ 2          │ 1       │ 1       │ -        │ -       │
└──────────────────────────────────────────────────────────────────────────────┴────────────┴─────────┴─────────┴──────────┴─────────┘[0m
[0m
[0m

 Tests exited with error 
"
`;
